exchanges/bittrex/bittrex_websocket.go

```diff
@@ -0,0 +1,627 @@
+	invocationIDCounter++
+	request := WsEventRequest{
+		Hub:          "c3",
+		Method:       "subscribe",
+		InvocationID: invocationIDCounter,
+	}
+
+	var channels []string
+	for i := range channelsToSubscribe {
+		channels = append(channels, channelsToSubscribe[i].Channel)
+	}
+	arguments := make([][]string, 0)
+	arguments = append(arguments, channels)
+	request.Arguments = arguments
+
+	requestString, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(request)
+	if err != nil {
+		return err
+	}
+
+	b.WsPendingRequests[request.InvocationID] = WsPendingRequest{
+		request,
+		&channelsToSubscribe,
+	}
+
+	return nil
+}
+
+// Unsubscribe sends a websocket message to stop receiving data from the channel
+func (b *Bittrex) Unsubscribe(channelsToUnsubscribe []stream.ChannelSubscription) error {
+	invocationIDCounter++
+	request := WsEventRequest{
+		Hub:          "c3",
+		Method:       "unsubscribe",
+		InvocationID: invocationIDCounter,
+	}
+
+	var channels []string
+	for i := range channelsToUnsubscribe {
+		channels = append(channels, channelsToUnsubscribe[i].Channel)
+	}
+	arguments := make([][]string, 0)
+	arguments = append(arguments, channels)
+	request.Arguments = arguments
+
+	requestString, err := json.Marshal(request)
+	if err != nil {
+		return err
+	}
+	if b.Verbose {
+		log.Debugf(log.ExchangeSys, "%s Sending JSON message - %s\n", b.Name, requestString)
+	}
+	err = b.Websocket.Conn.SendJSONMessage(request)
+	if err != nil {
+		return err
+	}
+
+	b.Websocket.RemoveSuccessfulUnsubscriptions(channelsToUnsubscribe...)
+	return nil
+}
+
+// wsReadData gets and passes on websocket messages for processing
+func (b *Bittrex) wsReadData() {
+	b.Websocket.Wg.Add(1)
+	defer b.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-b.Websocket.ShutdownC:
+			return
+		default:
+			resp := b.Websocket.Conn.ReadMessage()
+			if resp.Raw == nil {
+				log.Warnf(log.ExchangeSys, "%s Received empty message\n", b.Name)
+				return
+			}
+
+			err := b.wsHandleData(resp.Raw)
+			if err != nil {
+				b.Websocket.DataHandler <- err
+			}
+		}
+	}
+}
+
+func (b *Bittrex) wsDecodeMessage(encodedMessage string, v interface{}) error {
+	raw, err := crypto.Base64Decode(encodedMessage)
+	if err != nil {
+		return err
+	}
+	reader := flate.NewReader(bytes.NewBuffer(raw))
+	message, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(message, v)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Bittrex) wsHandleResponseData(request WsPendingRequest, respRaw []byte) error {
+	switch request.Method {
+	case "Authenticate":
+		var response WsAuthResponse
+		err := json.Unmarshal(respRaw, &response)
+		if err != nil {
+			log.Warnf(log.ExchangeSys, "%s - wsHandleResponseData - Cannot unmarshal into WsAuthResponse (%s)\n", b.Name, string(respRaw))
```

Comment:
If you can across just the websocket functionality change log.ExchangeSys -> log.WebsocketSys

Reply:
Updated as suggested (log.WebsocketMgr); note that there's no log.WebsocketSys (only log.RESTSys), and none of the websocket implementations seem to be consistent in their choice of loggers.

Reply:
Yeah, we need to be more consistent you are correct.